### PR TITLE
feat: Odd Even Sorting Algorithm

### PR DESCRIPTION
**Pull request by eddybruv**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C/pull/855"></a>

### CODE REVIEW

**Comment by eddybruv** [comment]
Please review and let me know if something is wrong @Panquesito7
Thank you!

File: sorting/odd_even_sort.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/**
+ * @file
+ * @author [Edwin Ajong] (https://github.com/eddybruv)
+ * @brief [Odd Even Sort] (https://en.wikipedia.org/wiki/Odd%E2%80%93even_sort) implementation
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:4]:
```suggestion
 * @author [Edwin Ajong](https://github.com/eddybruv)
 * @brief [Odd Even Sort](https://en.wikipedia.org/wiki/Odd%E2%80%93even_sort) implementation
```

File: sorting/odd_even_sort.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * @details
+ * This algorithm is divided into two phases- Odd and Even Phase.
+ * The algorithm runs until the array elements are sorted and in each iteration two phases occurs- Odd and Even Phases.
+ * In the odd phase, we perform a bubble sort on odd indexed elements and in the even phase,
+ * we perform a bubble sort on even indexed elements.
+ * Time Complexity: O(N ^ 2)
+ */
+
+#include <assert.h>     /// for assert
+#include <stdbool.h>    /// for boolean
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:14]:
```suggestion
#include <stdbool.h>    /// for bool
```

File: sorting/odd_even_sort.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            }
+        }
+    }
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+void test()
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:74]:
```suggestion
static void test()
```

File: sorting/odd_even_sort.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+
+int main(void)
+{
+    test();
+    return 0;
+}
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:104]:
```suggestion
}

```

File: sorting/odd_even_sort.c [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        assert(arr2[i] == arr2Soln[i]);    
+    }
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+
+int main(void)
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:100]:
```suggestion
int main()
```

File: sorting/odd_even_sort.c [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+{
+    test();
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:102]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/eddybruv/C/actions

**Comment by eddybruv** [comment]
> Great work! smile+1
> Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/eddybruv/C/actions

Thanks for the review @Panquesito7, I've done the changes that were requested.
Please review and let me know if there's anything wrong.

File: sorting/odd_even_sort.c [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+#include <stdlib.h>     /// for dynammic memory allocation
+#include <time.h>       /// for random number generation
+
+/**
+ * @brief Swap numbers by reference(using pointers)
+ * @param first pointer to first number
+ * @param second pointer to second number
+ * @returns void
+ */
+
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:25]:
```suggestion
```

File: sorting/odd_even_sort.c [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+void swap(int *first, int *second)
+{
+    int temp = *first;
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:28]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. **Requires adding the `inttypes.h` header.** Check other parts of the code ([reference](https://en.cppreference.com/w/c/types/integer)). 🙂

File: sorting/odd_even_sort.c [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+void oddEvenSort(int *arr, int size)
+{
+    bool isSorted = false;
+    while(!isSorted)
+    {
+        isSorted = true;
+        int i;
+
+        //Even phase
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:48]:
```suggestion
        // Even phase
```

File: sorting/odd_even_sort.c [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        for(i = 0; i <= size - 2; i += 2)
+        {
+            if(arr[i] > arr[i + 1])
+            {
+                swap(&arr[i], &arr[i + 1]);
+                isSorted = false;
+            }
+        }
+
+        //Odd phase
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:58]:
```suggestion
        // Odd phase
```

File: sorting/odd_even_sort.c [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        for(i = 1; i <= size - 2; i += 2)
+        {
+            if(arr[i] > arr[i + 1])
+            {
+                swap(&arr[i], &arr[i + 1]);
+                isSorted = false;
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:64]: Please add some documentation of what this code does here. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! Almost there! 😄

File: sorting/odd_even_sort.c (hunk 11 above)
Comment (eddybruv) [reply sorting/odd_even_sort.c]: @Panquesito7, I know the documentation maybe "amateur" but please it's my first time doing this, and it's quiet exiting.
Please take a look at what I've done and if there's any issue, I'll try my best to fix it.

File: sorting/odd_even_sort.c [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+ * is greater than the interger at the next array index (i.e arr[index + 2], to make sure the index is odd
+ * during the odd phase and even during the even phase).
+ * If the condition is true, the function "swap" is called and address of the intergers in question are passed as
+ * parameters. After the swap is completed, "isSorted" is set to "false".
+ * The while loop will keep running till the array is propertly sorted.
+ * @param arr array to be sorted
+ * @param size size of array
+ * @returns void
+ */
+
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:52]:
```suggestion
```

File: sorting/odd_even_sort.c [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+     for (int32_t i = 0; i < 4; i++)
+     {
+         assert(arr1[i] == arr1Soln[i]);    
+     }
+
+    for (int32_t i = 0; i < 10; i++)
+    {
+        assert(arr2[i] == arr2Soln[i]);    
+    }
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:104]: Add some documentation of what this code does here. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
It's just to address these reviews, and you're done! 😄

File: sorting/odd_even_sort.c (hunk 13 above)
Comment (eddybruv) [reply sorting/odd_even_sort.c:108]: Done as requested.

**Comment by eddybruv** [comment]
> 
> 
> It's just to address these reviews, and you're done! 😄

All done @Panquesito7

File: sorting/odd_even_sort.c [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                swap(&arr[i], &arr[i + 1]);
+                isSorted = false;
+            }
+        }
+    }
+}
+
+/**
+ * @brief Self-test implementations
+ * @details Two test(unsorted) arrays were created and their corresponding solution(sorted) arrays were also created.
```
Comment (Panquesito7) [inline sorting/odd_even_sort.c:84]:
```suggestion
 * @details Two tests (unsorted) arrays were created and their corresponding solution(sorted) arrays were also created.
```

**Review by Panquesito7 — DISMISSED** [review]
LGTM! Awesome work, @eddybruv! It's been great to be your first contribution here! We hope you keep contributing! 😄👍🎉
@amino19, this is a good opportunity to review a PR. 🙂

File: sorting/odd_even_sort.c [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+     for (int32_t i = 0; i < 4; i++)
+     {
+         assert(arr1[i] == arr1Soln[i]);    
+     }
+
+    for (int32_t i = 0; i < 10; i++)
+    {
+        assert(arr2[i] == arr2Soln[i]);    
+    }
+}
```
Comment (amino19) [inline sorting/odd_even_sort.c:109]: @eddybruv: Overall amazing work, just one line need to add, will confirm your tests are really passed!
```suggestion
    printf("All tests have passed!\n");
}
```

**Review by amino19 — CHANGES_REQUESTED** [review]

**Review by ayaankhan98 — DISMISSED** [review]
RSLGTM!

**Comment by eddybruv** [comment]
> LGTM! Awesome work, @eddybruv! It's been great to be your first contribution here! We hope you keep contributing! 😄👍🎉
> @amino19, this is a good opportunity to review a PR. 🙂
> 

It was an awesome experience, big thanks to @Panquesito7 for reviewing 🎉

**Review by Panquesito7 — APPROVED** [review]

File: sorting/odd_even_sort.c [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+     for (int32_t i = 0; i < 4; i++)
+     {
+         assert(arr1[i] == arr1Soln[i]);    
+     }
+
+    for (int32_t i = 0; i < 10; i++)
+    {
+        assert(arr2[i] == arr2Soln[i]);    
+    }
+    printf("All tests have passed!\n");
```
Comment (amino19) [inline sorting/odd_even_sort.c:109]: @eddybruv: LGTM! :partying_face:

**Review by amino19 — APPROVED** [review]

**Comment by Panquesito7** [comment]
Thanks for reviewing, @amino19 and @ayaankhan98! 😄

**Pull request merged**